### PR TITLE
[Request access] Add pings

### PR DESCRIPTION
**Pull request by erzhtor**
> **Warning**: Blocked by https://github.com/sourcegraph/analytics/pull/804

Closes  https://github.com/sourcegraph/sourcegraph/issues/47949.

This PR:
- Adds `access_requests_enabled` **critical telemetry** ping 
- Extends existing (**non-critical telemetry**) `growth_statistics` with the following new fields
  - `PendingAccessRequests`
  - `ApprovedAccessRequests`
  - `RejectedPendingRequests`

## Test plan
- Check diff
- Unit tests updated/added

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.11% (+3.14 kb) | 0.14% (+20.23 kb) 🔺 | 0.14% (+17.09 kb) 🔺 | 0.26% (+2) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/erzhtor/add-request-access-feature-pings/compare-report.html;tab=live_object) for a full comparison between the commits 8b953b1 and 3cb0cc6 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: internal/usagestats/growth.go [hunk 1]
```diff
@@ -64,11 +103,46 @@ SELECT COUNT(*) FILTER ( WHERE recent_usage_by_user.created_month = DATE_TRUNC('
...
+}
+
+func getAccessRequestsGrowthStatistics(ctx context.Context, db database.DB) (*accessRequestsGrowthStatistics, error) {
+	const accessRequestsQuery = `
+	SELECT
+		COUNT(*) FILTER (WHERE status LIKE 'PENDING') AS pending_access_requests,
+		COUNT(*) FILTER (WHERE status LIKE 'APPROVED') AS approved_access_requests,
+		COUNT(*) FILTER (WHERE status LIKE 'REJECTED') AS rejected_access_requests
+	FROM access_requests
+	WHERE DATE_TRUNC('month', created_at) = DATE_TRUNC('month', $1::timestamp)
```
Comment (mrnugget) [inline internal/usagestats/growth.go:128]: We don't have indexes on `created_at` and `status` -- I think we should at least add an index on `created_at`, right?

**Review by mrnugget — APPROVED** [review]

**Comment by rrhyne** [comment]
@erzhtor  can we also get ViewLoginPage, and LoginSuccessful? 

Having this information would help us understand how many times the login page is viewed, but a login does not occur.

File: cmd/frontend/internal/app/updatecheck/handler_test.go [hunk 2]
```diff
@@ -355,6 +358,90 @@ func TestSerializeBatchChangesUsage(t *testing.T) {
...
+		"has_external_url": "false",
+		"has_repos": "true",
+		"ever_searched": "false",
+		"ever_find_refs": "true",
+		"timestamp": "`+now.UTC().Format(time.RFC3339)+`"
+	}`)
+}
+
+func TestSerializeGrowthStatistics(t *testing.T) {
+	pr := &pingRequest{
```
Comment (kopancek) [inline cmd/frontend/internal/app/updatecheck/handler_test.go:373]: Probably out of scope of this PR, but this looks horrible to me - creating a struct with such amount of fields just to test deserialization. A better approach to me sounds like having a function that returns sane defaults and we can override what we need to override instead. Maybe a followup PR? 🤔

File: internal/usagestats/growth.go (hunk 1 above)
Comment (kopancek) [reply internal/usagestats/growth.go:128]: Should we also rip out `LIKE` comparison and use `=` operator instead? Would make it clearer to the naked eye that we are doing a direct string match.

Comment (erzhtor) [reply internal/usagestats/growth.go:128]: Good catch! Thank you for both suggestions. 1) Added indexes on the `created_at`, and `status` columns, 2) changed condition to use `create_at` index and 3) replaced `LIKE` with `=` operator ef20b342a36a807fa30aaf7020bdf265062e69a6.

File: cmd/frontend/internal/app/updatecheck/handler.go [hunk 3]
```diff
@@ -176,18 +176,19 @@ func canUpdateDate(clientVersionString string) (bool, error) {
...
-	ExternalServices     []string        `json:"extsvcs"`
-	BuiltinSignupAllowed bool            `json:"signup"`
-	HasExtURL            bool            `json:"hasExtURL"`
-	UniqueUsers          int32           `json:"u"`
-	Activity             json.RawMessage `json:"act"`
-	BatchChangesUsage    json.RawMessage `json:"batchChangesUsage"`
+	ClientSiteID          string `json:"site"`
+	LicenseKey            string
+	DeployType            string          `json:"deployType"`
+	ClientVersionString   string          `json:"version"`
```
[the remark this answers is not in the document]
Comment (erzhtor) [reply cmd/frontend/internal/app/updatecheck/handler.go]: Updated ac245b57c4a579ea9342e9b2c8e99f3c608acc7f.

File: cmd/frontend/internal/app/updatecheck/handler_test.go (hunk 2 above)
Comment (erzhtor) [reply cmd/frontend/internal/app/updatecheck/handler_test.go]: Good point! Updated ac245b57c4a579ea9342e9b2c8e99f3c608acc7f.

**Comment by erzhtor** [comment]
> @erzhtor can we also get ViewLoginPage, and LoginSuccessful?
> 
> Having this information would help us understand how many times the login page is viewed, but a login does not occur.

@rrhyne, we don't track ViewLoginPage, as events are tracked/allowed only for authenticated users. P.S.: This is probably smth to fix in future releases.

**Comment by rrhyne** [comment]
> @rrhyne, we don't track ViewLoginPage, as events are tracked/allowed only for authenticated users. P.S.: This is probably smth to fix in future releases.

@erzhtor I hear you but I'm hoping there is a simple solution we are not aware of, so I've asked the question in [#analytics](https://sourcegraph.slack.com/archives/CN4FC7XT4/p1677679484709969). Without this event we won't have the first stage of the funnel. 

I know we tracked it on the cloud previously via Amplitude, so fingers crossed.

**Comment by erzhtor** [comment]
> > @rrhyne, we don't track ViewLoginPage, as events are tracked/allowed only for authenticated users. P.S.: This is probably smth to fix in future releases.
> 
> @erzhtor I hear you but I'm hoping there is a simple solution we are not aware of, so I've asked the question in [#analytics](https://sourcegraph.slack.com/archives/CN4FC7XT4/p1677679484709969). Without this event we won't have the first stage of the funnel.
> 
> I know we tracked it on the cloud previously via Amplitude, so fingers crossed.

@rrhyne, in cloud (dotcom), it probably is working because we have a special mode when anonymous/unauthenticated users are allowed, which is turned off on the self-hosted version.

**Pull request merged**